### ISSUE TITLE
twisted.mail.test.test_mail.MaildirTestCase.testMailbox fails intermittently

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3812]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3812|
|Type|defect|
|Created|2009-05-03 15:20:31Z|
|Branch|https://github.com/twisted/twisted/tree/mailbox-order-3812|

trial -u produces a traceback in fairly short order:

```
[FAIL]: twisted.mail.test.test_mail.MaildirTestCase.testMailbox

Traceback (most recent call last):
  File "/home/glyph/Projects/Twisted/trunk/twisted/mail/test/test_mail.py", line 452, in testMailbox
    self.assertEquals(mb.listMessages(), range(1, 11))
twisted.trial.unittest.FailTest: not equal:
a = [4L, 5L, 6L, 7L, 8L, 9L, 10L, 1L, 2L, 3L]
b = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

<details><summary>Searchable metadata</summary>

```
trac-id__3812 3812
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_mailbox_order_3812 branches/mailbox-order-3812
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1241364031000000 1241364031000000
changetime__1241392683000000 1241392683000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26817]) Branching to 'mailbox-order-3812'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [26818]) Pad the microsecond field of the generated filenames so that the names sort properly

refs [#3812](https://github.com/twisted/twisted/issues/3812)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [26819]) fixup testMailbox

refs [#3812](https://github.com/twisted/twisted/issues/3812)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

This happened when microseconds value jumped up another digit (eg 7 to 10 or 96 to 100), causing files delivered later to have names which sorted earlier.  I've fixed this by padding the microsecond value to fix sorting.  A potential problem still remains - the seconds value will jump from 10 digits to 11 digits in 277 years; files created after that jump will sort before files created today, causing mailboxes to appear out of order.  I'm not going to fix that now, though.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/mailbox-order-3812)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @exarkun|
|-|-|

Hooray for private implementation details which can be incompatibly changed!

Land it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [26820]) Update copyright dates

refs [#3812](https://github.com/twisted/twisted/issues/3812)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [26821]) Merge mailbox-order-3812

Author: exarkun
Reviewer: glyph
Fixes: [#3812](https://github.com/twisted/twisted/issues/3812)

Fix an intermittent failure in the message ordering of the maildir-based
mailbox implementation.  Previously messages delivered within the same
second had a small chance to be delivered in such a way which that they
would appear to have been delivered in the opposite order as they really
were.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
